Framework/API/src/WorkspaceGroup.cpp

```diff
@@ -278,21 +279,43 @@ void WorkspaceGroup::workspaceDeleteHandle(
 }
 
 /**
- * Callback when a after-replace notification is received
- * Replaces a member if it was replaced in the ADS.
- * @param notice :: A pointer to a workspace after-replace notificiation object
+ * Callback when a before-replace notification is received
+ * Replaces a member if it was replaced in the ADS and checks
+ * for duplicate members within the group
+ * @param notice :: A pointer to a workspace before-replace notification object
  */
-void WorkspaceGroup::workspaceReplaceHandle(
+void WorkspaceGroup::workspaceBeforeReplaceHandle(
     Mantid::API::WorkspaceBeforeReplaceNotification_ptr notice) {
   std::lock_guard<std::recursive_mutex> _lock(m_mutex);
 
-  const std::string replacedName = notice->objectName();
-  for (auto &workspace : m_workspaces) {
-    if ((*workspace).name() == replacedName) {
-      workspace = notice->newObject();
+  const auto oldObject = notice->oldObject();
+  const auto newObject = notice->newObject();
+
+  bool foundOld(false);
+  bool foundDuplicate(false);
+
+  auto it = m_workspaces.begin();
```

Comment:
I have 1 final picky thing here. The scope of `it` can be reduced as it's not needed outside of the `for` loop:

`for (auto it = m_workspaces.begin(); it != m_workspaces.end(); ++it)`

This means that `it` cannot accidentally used outside of the loop.